src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs

```diff
@@ -54,6 +54,7 @@ public sealed class MultiSourcePackageMetadataProvider : IPackageMetadataProvide
 
         public async Task<IPackageSearchMetadata> GetPackageMetadataAsync(PackageIdentity identity, bool includePrerelease, CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Adding `ThrowIfCancellationRequested` calls is a mitigation rather than solving the real underlying problem. Could we instead try to pass cancellation tokens into more async methods like `MergeVersionsAsync`?

Reply:
Are you advocating to remove these and instead place them only in methods that don't call other rmethods?

How is this a mitigation? This method doesn't use the token, so saying that only the methods it calls should be responsible doesn't seem better to me.

The specific example you give has 2 lines and pretty minimal work each. Throwing between those 2 could be done, but it would be pretty rare that cancellation happened there.

Reply:
It's bad practice to not provide a `CancellationToken` to a method that accepts a `CancellationToken`, even if that `CancellationToken` parameter is optional. Unfortunately, NuGet has abstractions that lose the current `CancellationToken`, and as a result, NuGet does not propagate the `CancellationToken` to all methods. Once such abstraction is `MergeVersionsAsync`. Fixing these abstractions that lose the `CancellationToken` would be a more "correct" fix.

Reply:
I guess what I'm saying is, should we really accept/throw tokens from *every* method in our codebase? I think this example is too short to bother. The changes I actually did make are to our public API methods, and they contain several method calls. Those should have some checks, when possible

Reply:
> No, I'm saying that ThrowIfCancellationRequested is likely not the proper fix.

What is the proper fix?

Reply:
> MergeVersionsAsync does network requests without providing a CancellationToken:

I've not been this deep into Protocol before, but it looks to me that JSON is already retrieved and parsed at this point. If that's true, this merging is not making network calls...
https://github.com/NuGet/NuGet.Client/blob/bb60d6720d24890b8f3e071e70d27ea0f2bef57e/src/NuGet.Core/NuGet.Protocol/Model/PackageSearchMetadata.cs#L197

I throw before invoking this method, so if cancelled, this merge won't happen. All I can see that you're saying is if a token is cancelled between 2 items being merged, then it will eliminate subsequent merges.

Reply:
Say that you have the following C# code:

```
await httpClient.GetAsync("https://web-request-that-takes-10-seconds-to-respond");
await httpClient.GetAsync("https://another-web-request-that-takes-10-seconds-to-respond");
```

This hypothetical code takes 20 seconds, and maybe the user wants to cancel this. One way I can do this is by adding `ThrowIfCancellationRequested` on my cancellation token:

```
cancellationToken.ThrowIfCancellationRequested();
await httpClient.GetAsync("https://web-request-that-takes-10-seconds-to-respond");
cancellationToken.ThrowIfCancellationRequested();
await httpClient.GetAsync("https://another-web-request-that-takes-10-seconds-to-respond");
```

Now if I cancel during the first request, I don't have to wait for the second request... but if that first request is started, I need to wait for it to complete before cancellation kicks in! This isn't ideal. Here's another solution:

```
await httpClient.GetAsync("https://web-request-that-takes-10-seconds-to-respond", cancellationToken);
await httpClient.GetAsync("https://another-web-request-that-takes-10-seconds-to-respond", cancellationToken);
```

Now the customer can cancel the operation at any time, even if a 10 second web request has been started! The key takeaway here is that generally apps spends most of their time doing I/O (network requests, reading from disk, etc...). As a result, if you want to support cancellation properly, you need to cancel those I/O operations. Does that make sense?

Reply:
Could you acknowledge and reply to my questions here: https://github.com/NuGet/NuGet.Client/pull/3436#discussion_r437126351

Reply:
That's one implementation of `IPackageSearchMetadata.GetVersionsAsync` that seems to ok. Here's another implementation that is not:

https://github.com/NuGet/NuGet.Client/blob/0ba1a97052d9ba3f581bab5ab173270dac6681f0/src/NuGet.Core/NuGet.Protocol/Model/PackageSearchMetadataBuilder.cs#L51

Reply:
Loic, I'm declaring this out of scope for my current work, and created https://github.com/NuGet/Home/issues/9668